### PR TITLE
Add file ‘How to write a Koppeltaal connector’ for v1.2 (work in progress)

### PR DESCRIPTION
**Pull request by jbrokamp**
Pull request added for Koppeltaal documentation repository. Feedback expected in this commit thread.

- I put the document inside the "koppeltaal-1.2" folder in _docs.
- Configured a new category in the file: "How-to"
- Added an image in the image folder that should be linked in the document

### CODE REVIEW

File: _docs/koppelaal-1.2/write-connector.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
+---
+title: Write a Koppeltaal Connector
+category: How-to
+order: 2
+---
+
+# How-to: Koppeltaal connector
+
+## About this guide
+This guide will not go into Koppeltaal terminology, but rather assumes this knowledge is available to the reader. Whether it is from memory or the [Technical Design Document]. This document is not technical, but rather a descriptive / explanatory guide to assist the implementer towards a usable connector. Mainly by answering question in advance, that we think developers will bump into.
```
Comment (niels-bom-md) [inline _docs/koppelaal-1.2/write-connector.md]: answering question in advance ==> answering questions in advance

File: _docs/koppelaal-1.2/write-connector.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+Since the connector will be used by other parties and will also be handed over to the Koppeltaal Community at some point, it’s important that the source code is clear and readable for other developers that want to debug or add features in the (near) future.
+
+A git repository can be created at the [Koppeltaal organization on GitHub]. Sources of connectors in other programming languages can be viewed there as well.
+
+## Where to start
+- Find out which FHIR resources need to be communicated about. Application dependent.
+- Is there FHIR model library in my programming language?
+
+## Connector responsibilities
+A Koppeltaal connector is a library to be used by an application written in particular programming language. It communicates directly with the Koppeltaal Server REST interface, while exposing a simple API to the application, exchanging custom objects. Koppeltaal specific complexity should be handled by the connector - e.g., composing a message that conforms to Koppeltaal ontology.
```
Comment (niels-bom-md) [inline _docs/koppelaal-1.2/write-connector.md]: written in particular programming language ==> written in a particular programming language

File: _docs/koppelaal-1.2/write-connector.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+This guide will not go into Koppeltaal terminology, but rather assumes this knowledge is available to the reader. Whether it is from memory or the [Technical Design Document]. This document is not technical, but rather a descriptive / explanatory guide to assist the implementer towards a usable connector. Mainly by answering question in advance, that we think developers will bump into.
+
+## Connector repository and source code
+Since the connector will be used by other parties and will also be handed over to the Koppeltaal Community at some point, it’s important that the source code is clear and readable for other developers that want to debug or add features in the (near) future.
+
+A git repository can be created at the [Koppeltaal organization on GitHub]. Sources of connectors in other programming languages can be viewed there as well.
+
+## Where to start
+- Find out which FHIR resources need to be communicated about. Application dependent.
+- Is there FHIR model library in my programming language?
```
Comment (niels-bom-md) [inline _docs/koppelaal-1.2/write-connector.md]: Is there FHIR model library ==> Is there a FHIR model library

File: _docs/koppelaal-1.2/write-connector.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+- Retrieve the `Conformance Statement`
+- Send/retrieve `Messages` to/from a domain
+-- Create a Message bundle with FHIR Resources (send)
+-- Extract FHIR Resources from the Message bundle (receive)
+-- Update the _ProcessingStatus_ of a Message
+- Retrieve and update `Activity Definitions`
+
+Additionally, a connector could also support the following functionality:
+- Create a RESTful endpoint and configure a webhook on KT server
+- Using the `Storage Service` as data persistance utitliy
```
Comment (niels-bom-md) [inline _docs/koppelaal-1.2/write-connector.md]: persistance ==> persistence

File: _docs/koppelaal-1.2/write-connector.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+Obviously, the functionalities a connector should provide depends on the applications that use it. However, we think there is a set of basic functionalities that a connector in general should support.
+- Retrieve the `Conformance Statement`
+- Send/retrieve `Messages` to/from a domain
+-- Create a Message bundle with FHIR Resources (send)
+-- Extract FHIR Resources from the Message bundle (receive)
+-- Update the _ProcessingStatus_ of a Message
+- Retrieve and update `Activity Definitions`
+
+Additionally, a connector could also support the following functionality:
+- Create a RESTful endpoint and configure a webhook on KT server
```
Comment (niels-bom-md) [inline _docs/koppelaal-1.2/write-connector.md:39]: Is there documentation on the webhook already? Can we link to that?

File: _docs/koppelaal-1.2/write-connector.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+- Using the `Storage Service` as data persistance utitliy
+
+The connector will achieve most of this functionality by sending/retrieving Messages according to Koppeltaal specifications.
+view of the infrastructure entities that the connector will be dealing with.
```
Comment (niels-bom-md) [inline _docs/koppelaal-1.2/write-connector.md]: This sentence starts mid-way.

Comment (niels-bom-md) [reply _docs/koppelaal-1.2/write-connector.md]: Which functionality will a connector achieve outside of sending/retrieving Messages?  Maybe useful to add that?

File: _docs/koppelaal-1.2/write-connector.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+### Basic features
+
+#### The Conformance Statement
+A `Conformance Statement` is a set of capabilities of a FHIR Server. The conformance statement of the Koppeltaal Server contains information required for the OAuth2 implementation for the single sign on. Besides the capabilities, it contains a number of important endpoints that are needed for certain actions. It can be requested via _<Koppeltaal server url>/FHIR/Koppeltaal/metadata_ (e.g. http://edgekoppeltaal.vhscloud.nl/FHIR/Koppeltaal/metadata).
+The connector must extract these endpoints from the Conformance statement and use them for other actions. For instance, to request authorization for the launch of a game, the authorization endpoint must be read from the statement.
+
+#### Koppeltaal Messages
+Communication with the Koppeltaal server consists largely of exchanging `Messages`. The Koppeltaal message consists of a bundle of `FHIR Resources`. More information on the [FHIR resources] used in Koppeltaal.
```
Comment (niels-bom-md) [inline _docs/koppelaal-1.2/write-connector.md]: The Koppeltaal message ==> A Koppeltaal message (right?)

File: _docs/koppelaal-1.2/write-connector.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+The first resource in the bundle is the `MessageHeader`, which should always be present. The header contains important information about the content of the message, like the [message event types]. This information can be used to convert the resources in the bundle to a format that can be further processed by an application - e.g. an object holding all the important values to an application.
+The relevant values of the Resource should be copied to simple objects that can be handled by the application. It is possible that HL7 provides a library that provides FHIR models and validation schema's: [DSTU1] / [current version].
```
Comment (niels-bom-md) [inline _docs/koppelaal-1.2/write-connector.md:53]: When you say objects you don't necessarily mean programming objects right? We may have a connector that is written in a non-OO-language.

Also: shouldn't it mainly depend on the connector/programming language in what shape the Resource values are converted inside of the connector?

File: _docs/koppelaal-1.2/write-connector.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+_Note_: When building an outgoing message, mind that the MessageHeader should be the first Resource of the bundle, then add further resources.
+
+#### Setting Message ProcessingStatus
+It is important that the `ProcessingStatus` is changed when [handling messages]. That way the Koppeltaal server knows a message has been handled. Make sure to mark unprocessable messages as `Failed`, also providing an exception description.
+
+#### Activity Definitions
+`ActivityDefinitions` are treated somewhat different than the rest of the FHIR resources. They are of resource type `Other` and the content is as such not defined in the FHIR spec. In Koppeltaal, they should be created or updated without the use of Message resource bundles. A simple http POST request containing the [ActivityDefinition] in the request body suffices to create the ActivityDefinition.
```
Comment (niels-bom-md) [inline _docs/koppelaal-1.2/write-connector.md]: to create the ActivityDefinition ==> to create an ActivityDefinition

File: _docs/koppelaal-1.2/write-connector.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+-- In project root, to make it visible on GitHub
+-- How to use the connector? (examples)
+-- Technical details / caveats
+
+## Development process
+- Development process, unit / integration tests. 		Al doende. ‘praktisch’
+- Release management (versioning / compatibility) Development process - kern functionele dekking
+
+### Integration testing
+There is a domain specifically for connectors to execute their intergration tests with the Koppeltaal server. Som inspiration for tests:
```
Comment (niels-bom-md) [inline _docs/koppelaal-1.2/write-connector.md]: intergration ==> integration

Comment (niels-bom-md) [reply _docs/koppelaal-1.2/write-connector.md]: Som ==> Some

**Comment by niels-bom-md** [comment]
Good stuff!

I left some minor comments.

File: _docs/koppelaal-1.2/write-connector.md (hunk 5 above)
Comment (jbrokamp) [reply _docs/koppelaal-1.2/write-connector.md:39]: Unfortunately, I havent't found documentation on this yet. When it's there I will definitely link to that.

File: _docs/koppelaal-1.2/write-connector.md (hunk 6 above)
Comment (jbrokamp) [reply _docs/koppelaal-1.2/write-connector.md]: Working on it. Will be part of the pull request next time.

File: _docs/koppelaal-1.2/write-connector.md (hunk 8 above)
Comment (jbrokamp) [reply _docs/koppelaal-1.2/write-connector.md:53]: "e.g." means it's an example, it is indeed dependent on the programming language.

Also working out your last comment for the pull request next time.

**Review by jbrokamp — COMMENTED** [review]
Thanks for reviewing the doc, Niels!

**Comment by jbrokamp** [comment]
@relitodea I did another commit correcting the typo's, that @niels-bom-md  pointed out. Can you merge the pull request?

**Comment by jbrokamp** [comment]
@relitodea changed the branch to merge this pull request in to **develop**.

**Comment by relitodea** [comment]
Merged to master - after acceptance

**Pull request merged**